docs/source/traitsui_user_manual/testing.rst

```diff
@@ -0,0 +1,633 @@
+    with tester.create_ui(obj, dict(view=view)) as ui:
+        title_field = tester.find_by_name(ui, "title")
+        title_field.perform(KeySequence("New Title"))
+
+Note that there are limitations as to how truthful the GUI looks visually
+compared to manual testing.
+
+FAQ
+===
+
+.. rubric:: Is UITester GUI toolkit specific?
+
+It depends. The API of UITester is toolkit independent. As long as the behavior
+being tested is also toolkit independent, |UITester| should support test code
+with no toolkit specific logic. However if the underlying GUI behavior
+varies from toolkit to toolkit, the test code using |UITester| will need to
+adjust for that.
+
+.. rubric:: Which testing framework should I use with UITester?
+
+|UITester| does not depend on any testing framework. You should be able to use
+it with any testing framework (e.g. unittest, pytest).
+
+.. rubric:: Is UITester compatible with PyFace ModalDialogTester?
+
+Yes. For example, you can use |UITester| to launch a modal dialog, and use
+ModelDialogTester to close it.
+
+.. rubric:: Is UITester compatible with PyFace GuiTestAssistant?
+
+Yes.
+
+|UIWrapper.perform| and |UIWrapper.inspect| automatically request GUI
+events to be processed. Where they are used entirely for modifying GUI states,
+you may find uses of some of GuiTestAssistant features no longer necessary.
+
+.. rubric::
+   I use the delay parameter to play back my test but it does not look
+   identical to when I test the GUI manually, why?
+
+If the GUI / trait states being asserted in tests are not consistent compared
+to manually testing, then that is likely a bug. Please report it.
+
+If the GUI / trait states being asserted in tests are consistent with manual
+testing, then such visual discrepancies may have to be tolerated given there
+are toolkit-dependent and platform-dependent limitations as to what can be
+achieved for programmatically imitating user interactions.
+
+.. _advanced-testing:
+
+Advanced Testing
+================
+
+|UITester| supports testing TraitsUI editors with various user interaction
+logic. However it is possible that projects' test code may require
+additional logic that is not supported by |UITester| by default. Furthermore,
+some projects may implement and maintain their own custom UI editors. Those
+custom UI editors are also by default not supported by |UITester|.
+
+The API allows extension such that
+
+* projects can test TraitsUI editors with user interactions that do not
+  come supported by default.
+* projects can reuse the testing API for testing custom editors.
+
+.. note::
+   Extending support for testing a UI editor often requires knowledge of the
+   implementation details of the editor. If UI editor and the testing support
+   code are not maintained together, tests may be subject to breakages caused
+   by internal changes of the UI editors. With that, projects are encouraged
+   to contribute testing support upstream for testing UI editors they do not
```

Comment:
I think I was using the word upstream to mean something close to the context mentioned in this [article](https://en.wikipedia.org/wiki/Upstream_(software_development)). So in the context of an external project using TraitsUI, this sentence encourages developers of these projects to contribute test support extension back to TraitsUI so that it can be maintained here, and thereby reducing their tests' exposure to implementation details.

Maybe I should just remove this sentence. Contributing patches and codes upstream when you can is generally a good thing to do, so there is no need to repeat it here.

Reply:
That makes sense, I think it is perfectly fine if you want to leave it.  Its up to you